modules/tinymce/src/core/main/ts/api/EditorOptions.ts

```diff
@@ -0,0 +1,277 @@
+/**
+ * Copyright (c) Tiny Technologies, Inc. All rights reserved.
+ * Licensed under the LGPL or a commercial license.
+ * For LGPL see License.txt in the project root for license information.
+ * For commercial licenses see https://www.tiny.cloud/
+ */
+
+import { Arr, Fun, Obj, Optional, Strings, Type } from '@ephox/katamari';
+
+import Editor from './Editor';
+import { EditorOptions, NormalizedEditorOptions } from './OptionTypes';
+
+interface ProcessorSuccess<T> {
+  valid: true;
+  value: T;
+}
+
+interface ProcessorError {
+  valid: false;
+  message: string;
+}
+
+type SimpleProcessor<T> = (value: T) => boolean;
+type Processor<T, U> = (value: T) => ProcessorSuccess<U> | ProcessorError;
```

Comment:
```suggestion
type SimpleProcessor<T> = (value: unknown) => value is T;
type Processor<T> = (value: unknown) => ProcessorSuccess<T> | ProcessorError;
```

It just occurred to me that if these processors are trying to replace `Type.isString` or `Type.isNumber` then they would need to do that initial "is this the right type" check before they get to any other logic they might be looking at.

Reply:
This is incorrect and will cause issues with public usage, as the `value is T` won't work if you need to do this:
```
const isValid = (value: boolean) => value === false
```
So effectively this is **not** a type guard, it's just a regular predicate so the original is correct here.

Reply:
My issue is someone trying to write

```
const isMySecretKeyValid = (value: string) => value.length >= 32
```

and then passing in a number for the setting. We might not be able to make it fully a type guard, but it definitely should take `unknown` as an argument to hint to the users that they'll need to do some form of check

Reply:
Disagree very much there and passing `unknown` is going to cause a number of issues with integrators trying to use this API. In 99% of cases the type will be correct and this is the type of thing where it's external input so we need to assume the type is correct.

Reply:
I should add I do agree that `unknown` is technically the correct type here, don't forget we need to keep this easy to use for others. It's effectively the same as getting a server response whereby the you're just asserting the expected types as you're not going to type guard every little detail from the server response.

Reply:
I think the type of users who are enabling typescript and using advanced features of the TinyMCE API (like adding your own plugin with its own setting) are the types of users who would rather trade correctness in types for a little ease of use. Remember that they can just use plain JS, or `ts-ignore`, if its too much of a hassle. It would also make it easier on us developing our own plugins.

Reply:
I still disagree but I don't care enough to argue anymore so have made the change anyways.

Reply:
I didn't get a chance to say but I'm on Lee's side here. Although we could perhaps keep `unknown` and also alleviate the ease-of-use concerns by providing pre-built extensible processors that do a type guard and then run a custom function to validate.